### ISSUE TITLE
lazy datalodader

### ISSUE THREAD

**Issue by hhxxttxsh** [Victorwz/Open-Qwen2VL#17]
class LazyPretrainDataset(Dataset[Dict[str, torch.Tensor]]):
    def __init__(
        self,
        pretrain_pkl: Path,
        image_transform: ImageTransform,
        tokenizer: PreTrainedTokenizerBase,
        prompt_builder_fn: Type[PromptBuilder],
        train_num_samples: int,
        max_retries: int = 50,
        cache_size: Optional[int] = None,
    ) -> None:
        super().__init__()
        self.pretrain_pkl = pretrain_pkl
        self.image_transform = image_transform
        self.tokenizer = tokenizer
        self.prompt_builder_fn = prompt_builder_fn
        self.train_num_samples = train_num_samples
        self.dataset_type = "lazy-pretrain"
        
        # Lazy loading setup
        self.cache = {} if cache_size is None else {}
        self.cache_size = cache_size
        self.max_retries = max_retries
        
        # Store subset paths instead of all file paths
        self.subset_paths = self.pretrain_pkl.split(":")
        self._file_cache = {}  # Cache discovered files per path
        
        # Pre-compute total length for efficiency
        self._total_files = None

    def _get_files_for_path(self, subset_path: str) -> List[str]:
        """Lazy file discovery - only when needed"""
        if subset_path not in self._file_cache:
            print(f"Discovering files in: {subset_path}")
            files = glob.glob(os.path.join(subset_path, '**', '*.pkl'), recursive=True)
            self._file_cache[subset_path] = files
            print(f"Found {len(files)} files in {subset_path}")
        return self._file_cache[subset_path]

    def _get_file_path(self, idx: int) -> str:
        """Convert global index to actual file path"""
        current_idx = 0
        for subset_path in self.subset_paths:
            files = self._get_files_for_path(subset_path)
            if current_idx + len(files) > idx:
                file_idx = idx - current_idx
                return files[file_idx]
            current_idx += len(files)
        raise IndexError(f"Index {idx} out of range")

    def __len__(self) -> int:
        if self._total_files is None:
            self._total_files = sum(len(self._get_files_for_path(path)) for path in self.subset_paths)
        return self._total_files

    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
        """Get an item with retry logic and on-the-fly filtering"""
        for _ in range(self.max_retries):
            actual_idx = idx
            
            # Try to get from cache first
            if actual_idx in self.cache:
                return self.cache[actual_idx]
            
            # Load the sample
            sample = self._try_load_sample(actual_idx)
            
            if sample is not None:
                # Cache the sample if caching is enabled
                self._cache_sample(actual_idx, sample)
                return sample
            
            # If sample is invalid, try a random index
            actual_idx = random.randint(0, len(self) - 1)
        
        # If we couldn't find a valid sample after max_retries, return a dummy
        return self._get_dummy_sample()

    def _try_load_sample(self, idx: int) -> Optional[Dict[str, torch.Tensor]]:
        """Try to load a sample from the given index"""
        try:
            if idx >= len(self):
                return None
                
            file_path = self._get_file_path(idx)
            
            if file_path.endswith(".pkl"):
                with open(file_path, 'rb') as f:
                    data_pkl = pickle.load(f)
            elif file_path.endswith("packed.json"):
                with open(file_path, 'r') as f:
                    data_json = json.load(f)
                data_samples = [self.preprocess_obelics_interleaved(sample) for sample in data_json["data"]]
                data_samples = list(filter(None, data_samples))
                data_pkl = self.concat_document(data_samples, pad_token_id=self.tokenizer.pad_token_id)
            else:
                return None
            
            if data_pkl is None:
                return None
            
            # Process images
            image_tensors = [self.image_transform(image) for image in data_pkl['image_tensors']]
            data_pkl['image_tensors'] = image_tensors
            
            # Process labels
            labels = copy.deepcopy(data_pkl['input_ids'])
            labels[labels == IMAGE_TOKEN_INDEX] = IGNORE_INDEX
            data_pkl['labels'] = labels
            
            return data_pkl
            
        except Exception as e:
            print(f"Error loading sample at index {idx}: {e}")
            return None

    def _cache_sample(self, idx: int, sample: Dict[str, torch.Tensor]):
        """Cache a sample with LRU eviction if needed"""
        if self.cache_size is None or len(self.cache) < self.cache_size:
            self.cache[idx] = sample
        elif self.cache_size is not None and self.cache_size > 0:
            # Simple LRU: remove oldest item
            if self.cache:
                oldest_key = next(iter(self.cache))
                del self.cache[oldest_key]
            self.cache[idx] = sample

    def _get_dummy_sample(self) -> Dict[str, torch.Tensor]:
        """Return a dummy sample when no valid sample can be loaded"""
        # Create a minimal valid sample structure
        dummy_input_ids = torch.tensor([self.tokenizer.bos_token_id, self.tokenizer.eos_token_id])
        dummy_labels = torch.tensor([IGNORE_INDEX, IGNORE_INDEX])
        
        return {
            "image_tensors": [],
            "input_ids": dummy_input_ids,
            "labels": dummy_labels,
            "lengths": [2]
        }

    def get_modality_lengths(self) -> List[Tuple[bool, int]]:
        """Get a list of modalities (unimodal / text-only vs. multimodal) and length of conversations per example."""
        # For lazy loading, we can't pre-compute this efficiently
        # This would require loading all samples, which defeats the purpose
        print("Warning: get_modality_lengths() not supported for lazy loading datasets")
        return []

    def preprocess_obelics_interleaved(self, info):
        """
        Preprocess an interleaved image-text sequence, either by calling preprocess_gpt_interleaved (if the sequence
        is ChatGPT-generated) or by preprocessing in this function (if the sequences is from MMC4).
        """
        # image transform includes the padding process and normalization process
        
        texts = info["texts"]

        # load images first to find which ones are valid
        image_tensors, idxs = [], []
        for i, sample_image in enumerate(info["image_info"]):
            if not sample_image or "image_base64" not in sample_image:
                idxs.append(i)
                continue
            image_base64 = sample_image["image_base64"]
            rawbytes = base64.b64decode(image_base64)

            image = Image.open(io.BytesIO(rawbytes)).convert("RGB")
            if image.size[0] == 1 or image.size[1] == 1:
                continue

            image_tensors.append(image)
            idxs.append(i)

        if len(image_tensors) == 0 or len(image_tensors) >= 10:
            return None

        # preprocess and pad images
        texts = [texts[idx] for idx in idxs]
        if len(image_tensors) >= 2:
            texts =  [DEFAULT_IMAGE_TOKEN+"\n" if x is None else x.replace(DEFAULT_IMAGE_TOKEN, IMAGE_PLACEHOLDER)+END_CUNCK_TOKEN for x in texts]
        else:
            texts =  [DEFAULT_IMAGE_TOKEN+"\n" if x is None else x.replace(DEFAULT_IMAGE_TOKEN, IMAGE_PLACEHOLDER) for x in texts]
        corpus = "".join(texts)

        input_ids, labels = tokenizer_image_token(corpus, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors='pt')
        
        if input_ids[0] != self.tokenizer.bos_token_id:
            input_ids = torch.cat([torch.LongTensor([self.tokenizer.bos_token_id]), input_ids])

        assert len(image_tensors) == ((input_ids == IMAGE_TOKEN_INDEX).sum())
        
        # recompute the num of images
        final_num_images = (input_ids == IMAGE_TOKEN_INDEX).sum().item()
        
        # if all the images are at the end, then the doc becomes text-only doc, discard it
        if final_num_images == 0:
            return None
        
        assert len(image_tensors[:final_num_images]) == final_num_images
        
        if input_ids[-1] != self.tokenizer.eos_token_id:
            input_ids = torch.cat([input_ids, torch.LongTensor([self.tokenizer.eos_token_id])])
        
        # assert bos token
        assert input_ids[0] == self.tokenizer.bos_token_id

        length = (NUM_IMAGES_TOKEN - 1) * len(image_tensors) + input_ids.shape[-1]

        return (image_tensors[:final_num_images], input_ids, length)

    def concat_document(self, documents, pad_token_id, context_len=4096):
        new_input_ids = []
        new_image_tensors = []
        new_length = []
        for (image_tensors, input_ids, length) in documents:
            new_image_tensors += image_tensors
            new_input_ids.append(input_ids)
            new_length.append(length)

        new_input_ids = torch.cat(new_input_ids)

        if len(new_image_tensors) == 0:
            return None
        
        assert (new_input_ids == IMAGE_TOKEN_INDEX).sum() == len(new_image_tensors)
        assert (sum(new_length) - new_input_ids.shape[-1]) == (NUM_IMAGES_TOKEN - 1) * len(new_image_tensors)

        if new_input_ids.shape[-1] < context_len:
            new_input_ids = torch.cat([new_input_ids, torch.full((context_len - new_input_ids.shape[-1],), pad_token_id, dtype=new_input_ids.dtype)])

        return {"image_tensors": new_image_tensors, "input_ids": new_input_ids, "lengths": new_length}
